README.md
# WikiExtractor
[WikiExtractor.py](http://medialab.di.unipi.it/wiki/Wikipedia_Extractor) is a Python script that extracts and cleans text from a [Wikipedia database dump](http://download.wikimedia.org/).

The tool is written in Python and requires Python 2.7 or Python 3.3+ but no additional library.

For further information, see the [project Home Page](http://medialab.di.unipi.it/wiki/Wikipedia_Extractor) or the [Wiki](https://github.com/attardi/wikiextractor/wiki).

# Wikipedia Cirrus Extractor

`cirrus-extractor.py` is a version of the script that performs extraction from a Wikipedia Cirrus dump.
Cirrus dumps contain text with already expanded templates.

Cirrus dumps are available at:
[cirrussearch](http://dumps.wikimedia.org/other/cirrussearch/).

# Details

WikiExtractor performs template expansion by preprocessing the whole dump and extracting template definitions.

In order to speed up processing:

- multiprocessing is used for dealing with articles in parallel
- a cache is kept of parsed templates (only useful for repeated extractions).

## Installation

The script may be invoked directly, however it can be installed by doing:

    (sudo) python setup.py install

## Usage
The script is invoked with a Wikipedia dump file as an argument.
The output is stored in several files of similar size in a given directory.
Each file will contains several documents in this [document format](http://medialab.di.unipi.it/wiki/Document_Format).

    usage: WikiExtractor.py [-h] [-o OUTPUT] [-b n[KMG]] [-c] [--json] [--html]
                            [-l] [-s] [--lists] [-ns ns1,ns2]
                            [--templates TEMPLATES] [--no-templates] [-r]
                            [--min_text_length MIN_TEXT_LENGTH]
                            [--filter_disambig_pages] [-it abbr,b,big]
                            [-de gallery,timeline,noinclude] [--keep_tables]
                            [--processes PROCESSES] [-q] [--debug] [-a] [-v]
                            input

    Wikipedia Extractor:
    Extracts and cleans text from a Wikipedia database dump and stores output in a
    number of files of similar size in a given directory.
    Each file will contain several documents in the format:

        <doc id="" revid="" url="" title="">
            ...
            </doc>

    If the program is invoked with the --json flag, then each file will
    contain several documents formatted as json ojects, one per line, with
    the following structure

        {"id": "", "revid": "", "url":"", "title": "", "text": "..."}

    Template expansion requires preprocesssng first the whole dump and
    collecting template definitions.

    positional arguments:
      input                 XML wiki dump file

    optional arguments:
      -h, --help            show this help message and exit
      --processes PROCESSES
                            Number of processes to use (default 1)

    Output:
      -o OUTPUT, --output OUTPUT
                            directory for extracted files (or '-' for dumping to
                            stdout)
      -b n[KMG], --bytes n[KMG]
                            maximum bytes per output file (default 1M)
      -c, --compress        compress output files using bzip
      --json                write output in json format instead of the default one

    Processing:
      --html                produce HTML output, subsumes --links
      -l, --links           preserve links
      -s, --sections        preserve sections
      --lists               preserve lists
      -ns ns1,ns2, --namespaces ns1,ns2
                            accepted namespaces in links
      --templates TEMPLATES
                            use or create file containing templates
      --no-templates        Do not expand templates
      -r, --revision        Include the document revision id (default=False)
      --min_text_length MIN_TEXT_LENGTH
                            Minimum expanded text length required to write
                            document (default=0)
      --filter_disambig_pages
                            Remove pages from output that contain disabmiguation
                            markup (default=False)
      -it abbr,b,big, --ignored_tags abbr,b,big
                            comma separated list of tags that will be dropped,
                            keeping their content
      -de gallery,timeline,noinclude, --discard_elements gallery,timeline,noinclude
                            comma separated list of elements that will be removed
                            from the article text
      --keep_tables         Preserve tables in the output article text
                            (default=False)

    Special:
      -q, --quiet           suppress reporting progress info
      --debug               print debug info
      -a, --article         analyze a file containing a single article (debug
                            option)
      -v, --version         print program version


Saving templates to a file will speed up performing extraction the next time,
assuming template definitions have not changed.

Option --no-templates significantly speeds up the extractor, avoiding the cost
of expanding [MediaWiki templates](https://www.mediawiki.org/wiki/Help:Templates).

For further information, visit [the documentation](http://attardi.github.io/wikiextractor).


ChangeLog
2017-01-03  Michal Svamberg  <svamberg@civ.zcu.cz>

	* makehtmlfiles.pl: add new script for building web browsed content
	This script build simple pages from output of WikiExtractor.

	For more information read example section in README.md

	* WikiExtractor.py (keepPage): check for -xns option (or default = 0)
	* WikiExtractor.py (replaceInternalLinks): 
	  - Fix URL links to other pages
	  - Add support for Image links (make html tag <img src="%s" alt="%s">)
	* WikiExtractor.py (compact): add lines beginned by space into <tt> tag
	* WikiExtractor.py (pages_from): insert redirect pages if option
	--redirect_insert is enabled
	* WikiExtractor.py (main): add new options used in above changes
	  --all_links print links to all namespaces (independent of -xns or -ns)
	  --xml_namespaces add numeric namespaces for search pages
	  --redirect_insert create pages where used #REDIRECT wiki command

2017-01-15  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (process_dump): use text_type to decide whether
	to use decode('utf-8').)

2016-10-29  Giuseppe Attardi  <attardi@di.unipi.it>

	* setup.py: use scripts instead of console_scripts.

	* WikiExtractor.py (pages_from): handle self closing <text />.

2016-08-31  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (findMatchingBraces): fixed ambiguous cases
	like {{{!}} {{!}}}

2016-08-30  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (Extractor.wiki2text): drop templates before
	dropping tables.
	(Extractor.templateParams): fixed RE for separating var from value.

	(main): removed option xml_namespaces, since we extract only
	articles, i.e. those with namespace 0.

	(Extractor.clean): removed option --escapedoc, use --html instead.

	* WikiExtractor.py (Extractor.transform): Keep <nowiki> or else
	further expansion will be triggered.

2016-08-29  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (callParserFunction): reworked order of
	evaluation of template parameters, according to
	   https://www.mediawiki.org/wiki/Help:Templates#Order_of_evaluation
        Template parameters are fully evaluated before they are passed to the template.
	The branching parser function are an exception, since evaluation
	depends on values tested.

	* WikiExtractor.py (callParserFunction): reworked handling of
	parser function.
	Handle expansion of:
	   {{numero romano|1999}}
	where Template:Numero romano = {{{{{|safesubst:}}}#invoke:Numero romano|main}}
	and parameter body has no arguments, by getting them from frame.
	Instead expansion of:
	   {{Str endswith|1999|a.C.}}
	where Template:Str endswith = {{#ifeq:{{#invoke:String|sub|s={{{1|}}}| -{{#invoke:String|len|s={{{2}}}}} |ignore_errors=true}}|{{{2|}}}|sì}}
	uses parameters in template.

	* WikiExtractor.py (Extractor.extract): Adds an h1 tag for title if --html is
	passed. Patch contributed by okb1100 on GitHub.

	* WikiExtractor.py (modules): added partial emulation of modules String and Roman.
	)
2016-08-26  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (nowiki): handle <nowiki>...</nowiki>.

2016-08-22  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (Extractor.extract): fixed UTF-8 encoding with
	option -a.

2016-08-19  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (Extractor): fixed misspelled keepLists.

2016-08-11  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (Extract): made escape_doc a class variable.

2016-06-20  Seth Cleveland <scleveland@turnitin.com>

	* WikiExtractor.py: Fix string encoding error for python 2 and log
	extract exceptions.

2016-06-19  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py: support for Python 3 by orangain@gmail.com

2016-03-23  Seth Cleveland <scleveland@turnitin.com>

	* WikiExtractor.py: add filtering options -- min text length, xml
	namespace, and disambig pages. Added option to print the revision
	id with each document.

2016-03-23  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (compact): properly emit section headers when Extractor.keepSections.

2016-03-19  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (main): restored option --sections.

2016-03-12  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (findBalanced): fix to bad single tuple
	parameter to findBalanced.

2016-03-06  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (process_dump): handle templates with stdin.

2016-02-20  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (reduce_process): tell mapper to wait when
	spool gets too long.

2016-02-19  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (extract_process): use out.truncate() instread
	of out.close()

2016-02-15  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (Extractor.clean): turned into method.
	(process_dump): control spool length to avoid filing up memory.

2016-02-14  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (load_templates): save also <id> in templates.
	(pages_from): common iterator for extracting pages from dump, used
	both for analyzing pages, templates and single articles.

2016-02-13  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (reduce_process): close file here.

	* extractPage.py (process_data): allow range of ids.

2016-02-12  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (reduce_process): moved here creation of OutputSplitter.
	(compact): Extractor.keepLists allows preserving lists in output.
	(main): added new option --lists for preserving lists in output.
	(compact): rest lislLevel entering new section.
	(ignoredTags): removed 'div' from here, since it is in discardedTags.
	(ignoredTags): moved 'sub' and 'sup' to discardedTags.

2016-02-11  Giuseppe Attardi  <attardi@di.unipi.it>

	* WikiExtractor.py (if_empty): added implementation of Lua module
	for "Template:If empty".
	(compact): preserve lists also in text mode.
	(process_dump): log ID and title of pages.


cirrus-extract.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# =============================================================================
#  Version: 1.00 (December 15, 2015)
#  Author: Giuseppe Attardi (attardi@di.unipi.it), University of Pisa
#
# =============================================================================
#  Copyright (c) 2015. Giuseppe Attardi (attardi@di.unipi.it).
# =============================================================================
#  This file is part of Tanl.
#
#  Tanl is free software; you can redistribute it and/or modify it
#  under the terms of the GNU General Public License, version 3,
#  as published by the Free Software Foundation.
#
#  Tanl is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
# =============================================================================

"""Wikipedia Cirrus Extractor:
Extracts and cleans text from a Wikipedia Cirrus dump and stores output in a
number of files of similar size in a given directory.
Each file will contain several documents in the format:

	<doc id="" url="" title="">
        ...
        </doc>

"""

import sys, os.path, time
import re
import json
import argparse
import bz2
import gzip
import logging

# Program version
version = '1.00'

urlbase = 'http://it.wikipedia.org/'

# ----------------------------------------------------------------------

class NextFile(object):
    """
    Synchronous generation of next available file name.
    """

    filesPerDir = 100

    def __init__(self, path_name):
        self.path_name = path_name
        self.dir_index = -1
        self.file_index = -1

    def next(self):
        self.file_index = (self.file_index + 1) % NextFile.filesPerDir
        if self.file_index == 0:
            self.dir_index += 1
        dirname = self._dirname()
        if not os.path.isdir(dirname):
            os.makedirs(dirname)
        return self._filepath()

    def _dirname(self):
        char1 = self.dir_index % 26
        char2 = self.dir_index / 26 % 26
        return os.path.join(self.path_name, '%c%c' % (ord('A') + char2, ord('A') + char1))

    def _filepath(self):
        return '%s/wiki_%02d' % (self._dirname(), self.file_index)

class OutputSplitter(object):
    """
    File-like object, that splits output to multiple files of a given max size.
    """

    def __init__(self, nextFile, max_file_size=0, compress=True):
        """
        :param nextfile: a NextFile object from which to obtain filenames
            to use.
        :param max_file_size: the maximum size of each file.
        :para compress: whether to write data with bzip compression.
        """
        self.nextFile = nextFile
        self.compress = compress
        self.max_file_size = max_file_size
        self.file = self.open(self.nextFile.next())

    def reserve(self, size):
        if self.file.tell() + size > self.max_file_size:
            self.close()
            self.file = self.open(self.nextFile.next())

    def write(self, data):
        self.reserve(len(data))
        self.file.write(data)

    def close(self):
        self.file.close()

    def open(self, filename):
        if self.compress:
            return bz2.BZ2File(filename + '.bz2', 'w')
        else:
            return open(filename, 'w')

# ----------------------------------------------------------------------

class Extractor(object):

    def extract(self, out):
        """
        :param out: output file.
        """
        logging.debug("%s\t%s", self.id, self.title)
        text = ''.join(self.page)
        url = get_url(self.id)
        header = '<doc id="%s" url="%s" title="%s">\n' % (self.id, url, self.title)
        # Separate header from text with a newline.
        header += self.title + '\n\n'
        header = header.encode('utf-8')
        footer = "\n</doc>\n"
        out.write(header)
        text = clean(self, text)
        for line in compact(text):
            out.write(line.encode('utf-8'))
            out.write('\n')
        out.write(footer)

def process_dump(input_file, out_file, file_size, file_compress):
    """
    :param input_file: name of the wikipedia dump file; '-' to read from stdin
    :param out_file: directory where to store extracted data, or '-' for stdout
    :param file_size: max size of each extracted file, or None for no max (one file)
    :param file_compress: whether to compress files with bzip.
    """

    if input_file == '-':
        input = sys.stdin
    else:
        input = gzip.open(input_file)

    if out_file == '-':
        output = sys.stdout
        if file_compress:
            logging.warn("writing to stdout, so no output compression (use external tool)")
    else:
        nextFile = NextFile(out_file)
        output = OutputSplitter(nextFile, file_size, file_compress)

    # process dump
    # format
    # {"index":{"_type":"page","_id":"3825914"}}
    # {"namespace":0,"title":TITLE,"timestamp":"2014-06-29T15:51:09Z","text":TEXT,...}
    while True:
        line = input.readline()
        if not line:
            break
        index = json.loads(line)
        content = json.loads(input.readline())
        type = index['index']['_type']
        id = index['index']['_id']
        if type == 'page' and content['namespace'] == 0:
            title = content['title']
            text = content['text']
            # drop references:
            # ^ The Penguin Dictionary
            text = re.sub(r'  \^ .*', '', text)
            url = urlbase + 'wiki?curid=' + id
            header = '<doc id="%s" url="%s" title="%s">\n' % (id, url, title)
            page = header + title + '\n\n' + text + '\n</doc>\n'
            output.write(page.encode('utf-8'))

# ----------------------------------------------------------------------

# Minimum size of output files
minFileSize = 200 * 1024

def main():
    parser = argparse.ArgumentParser(prog=os.path.basename(sys.argv[0]),
        formatter_class=argparse.RawDescriptionHelpFormatter,
                                     description=__doc__)
    parser.add_argument("input",
                        help="Cirrus Json wiki dump file")
    groupO = parser.add_argument_group('Output')
    groupO.add_argument("-o", "--output", default="text",
                        help="directory for extracted files (or '-' for dumping to stdin)")
    groupO.add_argument("-b", "--bytes", default="1M",
                        help="maximum bytes per output file (default %(default)s)",
                        metavar="n[KMG]")
    groupO.add_argument("-c", "--compress", action="store_true",
                        help="compress output files using bzip")

    groupP = parser.add_argument_group('Processing')
    groupP.add_argument("-ns", "--namespaces", default="", metavar="ns1,ns2",
                        help="accepted namespaces")

    groupS = parser.add_argument_group('Special')
    groupS.add_argument("-q", "--quiet", action="store_true",
                        help="suppress reporting progress info")
    groupS.add_argument("-v", "--version", action="version",
                        version='%(prog)s ' + version,
                        help="print program version")

    args = parser.parse_args()

    try:
        power = 'kmg'.find(args.bytes[-1].lower()) + 1
        file_size = int(args.bytes[:-1]) * 1024 ** power
        if file_size < minFileSize:
            raise ValueError()
    except ValueError:
        logging.error('Insufficient or invalid size: %s', args.bytes)
        return

    FORMAT = '%(levelname)s: %(message)s'
    logging.basicConfig(format=FORMAT)

    logger = logging.getLogger()
    if not args.quiet:
        logger.setLevel(logging.INFO)

    input_file = args.input

    output_path = args.output
    if output_path != '-' and not os.path.isdir(output_path):
        try:
            os.makedirs(output_path)
        except:
            logging.error('Could not create: %s', output_path)
            return

    process_dump(input_file, output_path, file_size, args.compress)


if __name__ == '__main__':
    main()
